### PR TITLE
GEOMESA-2907 Remove Arrow Memory leak in ComplexArrowManualVisitor

### PR DESCRIPTION
**Pull request by jnh5y**
* Exposes ArrowAllocator.getAllocatedMemory and getPeakMemoryAllocation
methods.
* Addresses memory leak in ComplexArrowManualVisitor

Signed-off-by: Jim Hughes <jnh5y@ccri.com>

### CODE REVIEW

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileWriter.scala [hunk 1]
```diff
@@ -24,6 +24,7 @@ import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
   * For writing simple features to an arrow file.
   *
   * Uses arrow streaming format (no footer).
+  * Closing the SimpleFeatureArrowFileWriter closes the given os.
```
Comment (jnh5y) [inline geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileWriter.scala:27]: @elahrvivaz do you believe this statement?  (I tried to trace through the indirection here, etc.)

File: geomesa-process/geomesa-process-vector/src/test/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcessTest.scala [hunk 2]
```diff
@@ -42,20 +44,30 @@ class ArrowConversionProcessTest extends Specification {
...
       val bytes = process.execute(collection, null, null, null, null, null, null, null, null, null).reduce(_ ++ _)
       WithClose(SimpleFeatureArrowFileReader.streaming(() => new ByteArrayInputStream(bytes))) { reader =>
         reader.sft mustEqual sft
         SelfClosingIterator(reader.features()).map(ScalaSimpleFeature.copy).toSeq must
             containTheSameElementsAs(features)
       }
+      ArrowAllocator.getAllocatedMemory mustEqual 0
+    }
+
+    "encode a generic empty feature collection with dictionary values without leaking memory" in {
```
Comment (jnh5y) [inline geomesa-process/geomesa-process-vector/src/test/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcessTest.scala:67]: This case was the crux of this smaller, corner case memory leak.

File: geomesa-process/geomesa-process-vector/src/test/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcessTest.scala [hunk 3]
```diff
@@ -42,20 +44,30 @@ class ArrowConversionProcessTest extends Specification {
...
+      // This returns an empty iterator.
+      process.execute(new ListFeatureCollection(sft), null, null, null, Seq("name"), null, null, null, null, null)
+      ArrowAllocator.getAllocatedMemory mustEqual 0
```
Comment (jnh5y) [inline geomesa-process/geomesa-process-vector/src/test/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcessTest.scala:70]: @elahrvivaz I'm curious if you can suggest places where we could test for this kind of allocation in unit tests where queries time out....

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileWriter.scala (hunk 1 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileWriter.scala:27]: tracing through it looks like yes, although there appears to be a small possibility that an exception during the close would cause it to not be invoked: https://github.com/apache/arrow/blob/master/java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowWriter.java#L202

Comment (jnh5y) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileWriter.scala:27]: Given that, would it be better to use a WithClose around the allocation of the outputstream as well?  Would closing the outstream multiple times cause trouble?

File: geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcess.scala [hunk 4]
```diff
@@ -262,25 +262,28 @@ object ArrowConversionProcess {
...
-          } else {
-            CloseWithLogging(writer)
-          }
-          out.toByteArray
+          writer.flush()
+          bytes.append(out.toByteArray)
+          out.reset()
         }
+
+        bytes.append(out.toByteArray)
```
Comment (elahrvivaz) [inline geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcess.scala:285]: i think these 2 lines (and the declaration of `bytes`) should be moved outside the WithClose block, so that any footer bytes written by the SimpleFeatureWriter get returned correctly. It might not matter if the footer is missing, but seems safer.

File: geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcess.scala [hunk 5]
```diff
@@ -262,25 +262,28 @@ object ArrowConversionProcess {
         case Some(o) => features.sorted(o).iterator
       }
 
+      buildResult(dictionaries, sorted)
+    }
+
+    private def buildResult(dictionaries: Map[String, ArrowDictionary], sorted: Iterator[SimpleFeature]) = {
```
Comment (elahrvivaz) [inline geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcess.scala:268]: could you add the return type to the method signature?

File: geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcess.scala (hunk 4 above)
Comment (jnh5y) [reply geomesa-process/geomesa-process-vector/src/main/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcess.scala:285]: Hmm... if ```bytes.append(out.toByteArray)``` is moved outside of the WithClose block, wouldn't ```out``` be closed by the WithClose?

File: geomesa-process/geomesa-process-vector/src/test/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcessTest.scala [hunk 6]
```diff
@@ -42,20 +44,30 @@ class ArrowConversionProcessTest extends Specification {
 
   "ArrowConversionProcess" should {
     "encode an empty feature collection" in {
+      ArrowAllocator.getAllocatedMemory mustEqual 0
```
Comment (elahrvivaz) [inline geomesa-process/geomesa-process-vector/src/test/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcessTest.scala:47]: hmm, i don't think this is a safe check - even with `sequential`, the root allocator is shared across all unit tests, which often run in parallel. We could maybe modify the process method signature or constructor to pass in an optional allocator so that we can check it?

Comment (jnh5y) [reply geomesa-process/geomesa-process-vector/src/test/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcessTest.scala:47]: Good eye!

Would we need to thread the optional allocator through to the SimpleFeatureVector / Writer classes?  (Or is that mostly already plumbed?)

File: geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileWriter.scala (hunk 1 above)
Comment (elahrvivaz) [reply geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/SimpleFeatureArrowFileWriter.scala:27]: hmm, generally i don't think closing multiple times would be an issue, but not sure if that's 100% true. probably any exceptions during the close would be infrequent and not result in massive memory leaks so maybe it's ok as is.

**Review by elahrvivaz — APPROVED** [review]

File: geomesa-process/geomesa-process-vector/src/test/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcessTest.scala [hunk 7]
```diff
@@ -58,6 +60,12 @@ class ArrowConversionProcessTest extends Specification {
       }
     }
 
+    "encode a generic empty feature collection with dictionary values without leaking memory" in {
+      // This returns an empty iterator.
+      process.execute(new ListFeatureCollection(sft), null, null, null, Seq("name"), null, null, null, null, null)
+      ArrowAllocator.getAllocatedMemory(sft.getTypeName) must equalTo(0)
```
Comment (elahrvivaz) [inline geomesa-process/geomesa-process-vector/src/test/scala/org/locationtech/geomesa/process/transform/ArrowConversionProcessTest.scala:66]: we should make the type name more unique to avoid interference from any other tests

**Review by elahrvivaz — APPROVED** [review]

**Comment by elahrvivaz** [comment]
fixed up as 42fe5a44751bfee9455d3267e937f5749ec0d175
